azurelinuxagent/ga/exthandlers.py

```diff
@@ -548,108 +561,173 @@ def _report_error_event_and_return_false(error_message):
+            handler_state = ext_handler.properties.state
 
             # The Guest Agent currently only supports 1 installed version per extension on the VM.
             # If the extension version is unregistered and the customers wants to uninstall the extension,
             # we should let it go through even if the installed version doesnt exist in Handler manifest (PIR) anymore.
             # If target state is enabled and version not found in manifest, do not process the extension.
-            if ext_handler_i.decide_version(target_state=state) is None and state == ExtensionRequestedState.Enabled:
+            if ext_handler_i.decide_version(target_state=handler_state,
+                                            extension=extension) is None and handler_state == ExtHandlerRequestedState.Enabled:
                 handler_version = ext_handler_i.ext_handler.properties.version
                 name = ext_handler_i.ext_handler.name
                 err_msg = "Unable to find version {0} in manifest for extension {1}".format(handler_version, name)
                 ext_handler_i.set_operation(WALAEventOperation.Download)
-                ext_handler_i.set_handler_status(message=ustr(err_msg), code=-1)
-                ext_handler_i.report_event(message=ustr(err_msg), is_success=False)
-                return False
-
-            ext_handler_i.logger.info("Target handler state: {0} [incarnation {1}]", state, etag)
-            if state == ExtensionRequestedState.Enabled:
-                self.handle_enable(ext_handler_i)
-            elif state == ExtensionRequestedState.Disabled:
-                self.handle_disable(ext_handler_i)
-            elif state == ExtensionRequestedState.Uninstall:
-                self.handle_uninstall(ext_handler_i)
+                raise ExtensionError(msg=err_msg)
+
+            # Handle everything on an extension level rather than Handler level
+            ext_handler_i.logger.info("Target handler state: {0} [incarnation {1}]", handler_state, etag)
+            if handler_state == ExtHandlerRequestedState.Enabled:
+                self.handle_enable(ext_handler_i, extension)
+            elif handler_state == ExtHandlerRequestedState.Disabled:
+                self.handle_disable(ext_handler_i, extension)
+            elif handler_state == ExtHandlerRequestedState.Uninstall:
+                self.handle_uninstall(ext_handler_i, extension=extension)
             else:
-                message = u"Unknown ext handler state:{0}".format(state)
+                message = u"Unknown ext handler state:{0}".format(handler_state)
                 raise ExtensionError(message)
 
             return True
-
+        except MultiConfigExtensionError as error:
+            ext_name = ext_handler_i.get_extension_full_name(extension)
+            err_msg = "Error processing MultiConfig extension {0}: {1}".format(ext_name, ustr(error))
+            # This error is only thrown for enable operation on MultiConfig extension.
```

Comment:
Should we rename it to MultiConfigExtensionEnableError to make this clearer and help avoid raising those on other operations?

Reply:
Makes sense. Initially I was raising this for both Disable and Enable errors but then added the code to ignore errors for disable, but renaming it makes sense. Let me do it